src/internal/connector/onedrive/url_cache.go

```diff
@@ -0,0 +1,275 @@
+		servicer graph.Servicer,
+		driveID, link string,
+	) itemPager
+
+	errors *fault.Bus
+}
+
+// newURLache creates a new URL cache for the specified drive ID
+func newURLCache(
+	driveID string,
+	refreshInterval time.Duration,
+	svc graph.Servicer,
+	itemPagerFunc func(
+		servicer graph.Servicer,
+		driveID, link string,
+	) itemPager,
+) (*urlCache, error) {
+	err := validateCacheParams(
+		driveID,
+		refreshInterval,
+		svc,
+		itemPagerFunc)
+	if err != nil {
+		return nil, clues.Wrap(err, "invalid cache parameters")
+	}
+
+	return &urlCache{
+			urlMap:          make(map[string]itemProperties),
+			lastRefreshTime: time.Time{},
+			driveID:         driveID,
+			refreshInterval: refreshInterval,
+			svc:             svc,
+			itemPagerFunc:   itemPagerFunc,
+			errors:          fault.New(false),
+		},
+		nil
+}
+
+// validateCacheParams validates the parameters passed to newURLCache
+func validateCacheParams(
+	driveID string,
+	refreshInterval time.Duration,
+	svc graph.Servicer,
+	itemPagerFunc func(
+		servicer graph.Servicer,
+		driveID, link string,
+	) itemPager,
+) error {
+	if len(driveID) == 0 {
+		return clues.New("drive id is empty")
+	}
+
+	if refreshInterval < 0 {
+		return clues.New("invalid refresh interval")
+	}
+
+	if svc == nil {
+		return clues.New("nil graph servicer")
+	}
+
+	if itemPagerFunc == nil {
+		return clues.New("nil item pager")
+	}
+
+	return nil
+}
+
+// getItemProps returns the item properties for the specified drive item ID
+func (uc *urlCache) getItemProperties(
+	ctx context.Context,
+	itemID string,
+) (*itemProperties, error) {
+	if len(itemID) == 0 {
+		return nil, clues.New("item id is empty")
+	}
+
+	ctx = clues.Add(ctx, "drive_id", uc.driveID)
+
+	// Lazy refresh
+	if uc.needsRefresh() {
+		err := uc.refreshCache(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	url, err := uc.readCache(ctx, itemID)
+	if err != nil {
+		return nil, err
+	}
+
+	return url, nil
+}
+
+// needsRefresh returns true if the cache is empty or if refresh interval has
+// elapsed
+func (uc *urlCache) needsRefresh() bool {
+	uc.cacheLock.RLock()
+	defer uc.cacheLock.RUnlock()
+
+	return len(uc.urlMap) == 0 || time.Since(uc.lastRefreshTime) > uc.refreshInterval
+}
+
+// refreshCache refreshes the URL cache by performing a delta query.
+func (uc *urlCache) refreshCache(
+	ctx context.Context,
+) error {
+	// Acquire mutex to prevent multiple threads from refreshing the
+	// cache at the same time
+	uc.refreshMutex.Lock()
+	defer uc.refreshMutex.Unlock()
+
+	// If the cache was refreshed by another thread while we were waiting
+	// to acquire mutex, return
+	if !uc.needsRefresh() {
+		return nil
+	}
+
+	// Hold cache lock in write mode for the entire duration of the refresh.
+	// This is to prevent other threads from reading the cache while it is
+	// being updated page by page
+	uc.cacheLock.Lock()
+	defer uc.cacheLock.Unlock()
+
+	// Issue a delta query to graph
+	logger.Ctx(ctx).Debug("refreshing url cache")
```

Comment:
I think this is significant enough (and, more importantly, infrequent enough) to warrant an info-level.
```suggestion
	logger.Ctx(ctx).Info("refreshing url cache")
```

Also, these logs are effectively duplicates of the `deltaQuery` logs.  Better off just to let that func do the logging.

Reply:
> Also, these logs are effectively duplicates of the deltaQuery logs. Better off just to let that func do the logging.

I would still like to keep the refresh event logs to differentiate between normal delta queries vs delta queries issued by the cache.

Reply:
Misread your statement. I'll remove the logs in `deltaQuery`